File: example_contests/fk_2014_beta/problems/evil_odious/solution.py
mem = {}
def cnt(n, e):
    if n == 0:
        return 1 if e else 0

    if (n,e) in mem:
        return mem[(n,e)]

    res = cnt(n-1, e) + cnt(n-1, not e)
    mem[(n,e)] = res
    return res

def nth(n, k, e):
    if n == 0:
        return ""

    if k < cnt(n-1, e):
        return "0" + nth(n-1, k, e)
    else:
        return "1" + nth(n-1, k-cnt(n-1, e), not e)

def parse_bin(s):
    res = 0
    for b in s:
        res *= 2
        if b == '1':
            res += 1
    return res

n = int(input())
print(parse_bin(nth(100, n-1, True)) + parse_bin(nth(100, n-1, False)))

File: example_contests/fk_2014_beta/problems/evil_odious/solution.cpp
#include <algorithm>
#include <bitset>
#include <cassert>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <ctime>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <list>
#include <map>
#include <queue>
#include <set>
#include <sstream>
#include <stack>
#include <string>
#include <utility>
#include <vector>
using namespace std;

#define all(o) (o).begin(), (o).end()
#define allr(o) (o).rbegin(), (o).rend()
const int INF = 2147483647;
typedef long long ll;
typedef pair<int, int> ii;
typedef vector<int> vi;
typedef vector<ii> vii;
typedef vector<vi> vvi;
typedef vector<vii> vvii;
template <class T> int size(T &x) { return x.size(); }

// assert or gtfo

ll cnt[70][2];

ll nth(int bs, ll n, bool even)
{
    ll res = 0;
    for (int i = bs; i > 0; i--)
    {
        res <<= 1;

        if (n >= cnt[i-1][even])
        {
            n -= cnt[i-1][even];
            res |= 1;
            even = !even;
        }
    }

    return res;
}

int main()
{
    for (int i = 0; i < 70; i++)
    {
        for (int j = 0; j < 2; j++)
        {
            if (i == 0)
            {
                cnt[i][j] = j == 1 ? 1 : 0;
            }
            else
            {
                cnt[i][j] = cnt[i-1][j] + cnt[i-1][!j];
            }
        }
    }

    ll n;
    cin >> n;
    n--;

    cout << nth(64, n, false) + nth(64, n, true) << endl;

    return 0;
}

